sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -3039,7 +3039,7 @@ object Encode {
       legacyCharsets: Boolean,
       legacyErrorAction: Boolean): Array[Byte] = {
     val toCharset = charset.toString
-    if (input.numBytes == 0 || "UTF-8".equalsIgnoreCase(toCharset)) {
+    if (input.numBytes == 0 || ("UTF-8".equalsIgnoreCase(toCharset) && input.isValid())) {
```

Comment:
can we also add a test for invalid utf8 input?

Reply:
Ok, sure

Reply:
BTW let's also test empty bytes. Did we fail before this optimization?

Reply:
I can't create a 'failed case before or after this PR. Invalid inputs do not always generate errors, but a re-mapped byte array. It's tricky for me to find a specific case to test `CharacterCodingException`